algorithmic_efficiency/param_utils.py

```diff
@@ -1,38 +1,45 @@
 """Utilities for dealing with parameter-related logic like types and shapes."""
+import flax
+import jax
+
 from algorithmic_efficiency import spec
 
 
+def pytorch_param_shapes(model):
+  return {k: spec.ShapeTuple(v.shape) for k, v in model.named_parameters()}
+
+
 def pytorch_param_types(param_shapes):
-  if param_shapes is None:
-    raise ValueError(
-        'This should not happen, workload.init_model_fn() should be called '
-        'before workload.model_params_types!')
   param_types = {}
   for name in param_shapes.keys():
     if 'bias' in name:
       param_types[name] = spec.ParameterType.BIAS
-    elif 'BatchNorm' in name:
+    elif 'bn' in name:
       param_types[name] = spec.ParameterType.BATCH_NORM
-    elif 'Conv' in name:
+    elif 'conv' in name:
       param_types[name] = spec.ParameterType.CONV_WEIGHT
-    elif 'Embedding' in name:
+    elif 'embedding' in name:
       param_types[name] = spec.ParameterType.EMBEDDING
     else:
       param_types[name] = spec.ParameterType.WEIGHT
   return param_types
 
 
-def jax_param_types(param_tree):
+def jax_param_shapes(params):
+  return jax.tree_map(lambda x: spec.ShapeTuple(x.shape), params)
+
+
+def jax_param_types(param_shapes, parent_name=''):
   param_types_dict = {}
-  for name, value in param_tree.items():
-    if isinstance(value, dict):
-      param_types_dict[name] = jax_param_types(value)
+  for name, value in param_shapes.items():
+    if isinstance(value, dict) or isinstance(value, flax.core.FrozenDict):
+      param_types_dict[name] = jax_param_types(value, parent_name=name)
     else:
       if 'bias' in name:
         param_types_dict[name] = spec.ParameterType.BIAS
-      elif 'BatchNorm' in name:
+      elif 'BatchNorm' in parent_name:
         param_types_dict[name] = spec.ParameterType.BATCH_NORM
-      elif 'Conv' in name:
+      elif 'Conv' in parent_name:
```

Comment:
not sure if this matters, but maybe to be safer we should do case insensitive comparisons? like:
```python
      elif 'batchnorm' in parent_name.lower():
        param_types_dict[name] = spec.ParameterType.BATCH_NORM
      elif 'conv' in parent_name.lower():
```

Reply:
That's a good point. Fixed!